File: Awesome.AI.Source/Awesome.AI/Core/Spaces/HubSpace.cs
using Awesome.AI.CoreInternals;
using Awesome.AI.Variables;
using static Awesome.AI.Variables.Enums;

namespace Awesome.AI.Core.Spaces
{
    public class HubSpace
    {
        /*
         * these are groups of UNITs
         * */

        public Dictionary<string, double> weights { get; set; }

        private TheMind mind;
        public HubSpace(TheMind mind)
        {
            this.mind = mind;
            this.weights = new Dictionary<string, double>();

            MINDS mindtype = mind.mindtype;
            Lookup lookup = new Lookup();
            string[] occus = lookup.occupasions;

            Random rand = new Random();
            foreach (string occ in occus)
            {
                List<string> hubs = lookup.GetHUBS(mindtype, occ);
                foreach (string hub in hubs)
                {
                    if (weights.ContainsKey(hub))
                        continue;

                    double _r = rand.NextDouble();
                    weights.Add(hub, _r);
                }
            }
        }

        public void AdjustWeights(string sub, double _val)
        {
            try
            {
                string occu = mind._internal.Occu.name;
                MINDS mindtype = mind.mindtype;

                Lookup lookup = new Lookup();
                List<string> hubs = lookup.GetHUBS(mindtype, occu);

                double _val1 = _val;
                double _val2 = _val * 0.1d;

                foreach (string hub in hubs)
                    weights[hub] -= _val2;

                foreach (string hub in hubs)
                    weights[hub] = weights[hub] < 0.0d ? 0.0d : weights[hub]; 
                
                weights[sub] += _val1;
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        public double GetIndex(string sub_get)
        {
            try
            {
                string occu = mind._internal.Occu.name;
                MINDS mindtype = mind.mindtype;

                Lookup lookup = new Lookup();
                List<string> hubs = lookup.GetHUBS(mindtype, occu);

                var weights_occ = new List<KeyValuePair<string, double>>();
                hubs.ForEach(x => weights_occ.Add(KeyValuePair.Create(x, weights[x])));

                // no ordering, should be based on semantics
                var sum = weights_occ.Sum(x => x.Value);

                double res = 0.0d;
                double count = 0.0d;
                int i = 0;
                foreach (var weight in weights_occ)
                {
                    var _w = weight.Value;
                    var area = (_w / sum) * 100.0d;

                    count += area;

                    var sub = GetSubject(count);

                    if (sub_get == sub)
                        break;

                    var weight_next = weights_occ[i + 1].Value;
                    var area_next = (weight_next / sum) * 100.0d;

                    res = count + area_next / 2;
                    i++;
                }

                return res;
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        private bool LongDecision(UNIT unit, out string sub)
        {
            sub = "";
            if (unit.IsDECISION())
            {
                string data = unit.Data;

                // "SHOULD_A" "SHOULD_B" "SHOULD_C"
                if (data.StartsWith(CONST.lng_should))
                    sub = CONST.LSUB_SHOULD; 

                // "WHAT_KITCHEN" "WHAT_BEDROOM" "WHAT_LIVINGROOM" "WHAT_im busy right now.." "WHAT_not right now.." "WHAT_talk later.."                
                if (data.StartsWith(CONST.lng_what))
                    sub = CONST.LSUB_WHAT; 
            }

            return sub != "";
        }

        public string GetSubject(UNIT unit)
        {
            try
            {
                if (LongDecision(unit, out string sub))
                    return sub;

                string occu = mind._internal.Occu.name;

                if (occu == "init")
                    return "init";

                MINDS mindtype = mind.mindtype;

                Lookup lookup = new Lookup();
                List<string> hubs = lookup.GetHUBS(mindtype, occu);

                var weights_occ = new List<KeyValuePair<string, double>>();
                hubs.ForEach(x => weights_occ.Add(KeyValuePair.Create(x, weights[x])));

                // no ordering, should be based on semantics
                var sum = weights.Sum(x => x.Value);

                double count = 0.0d;
                foreach (var weight in weights_occ)
                {
                    var _w = weight.Value;
                    var area = (_w / sum) * 100.0d;

                    count += area;

                    sub = weight.Key;

                    if (count >= unit.HI)
                        break;
                }

                return sub;
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        public string GetSubject(double hub_index)
        {
            try
            {
                string occu = mind._internal.Occu.name;
                MINDS mindtype = mind.mindtype;

                Lookup lookup = new Lookup();
                List<string> hubs = lookup.GetHUBS(mindtype, occu);

                var weights_occ = new List<KeyValuePair<string, double>>();
                hubs.ForEach(x => weights_occ.Add(KeyValuePair.Create(x, weights[x])));

                // no ordering, should be based on semantics
                var sum = weights.Sum(x => x.Value);

                string sub = "";
                double count = 0.0d;
                foreach (var weight in weights_occ)
                {
                    var _w = weight.Value;
                    var area = (_w / sum) * 100.0d;

                    count += area;

                    sub = weight.Key;
                 
                    if (count >= hub_index)
                        break;
                }

                return sub;
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        public List<UNIT> UnitsPerHub(string hub)
        {
            try
            {
                var units = mind.access.UNITS_ALL();

                List<UNIT> res = new List<UNIT>();

                foreach (var unit in units)
                {
                    var sub = GetSubject(unit.HI);
                    if (sub == hub && !res.Contains(unit))
                        res.Add(unit);
                }

                return res;
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        public List<UNIT> UnitsPerOccupasionc()
        {
            try
            {
                //string occu = mind._internal.Occu.name;
                
                var units = mind.access.UNITS_ALL();
                
                List<UNIT> res = new List<UNIT>();

                foreach (var unit in units)
                {
                    //var max = Max(unit.register);
                    //if (max == occu && !res.Contains(unit))
                    //    res.Add(unit);

                    if (unit.IsValid && !res.Contains(unit))
                        res.Add(unit);
                }            
            
                return res;
            }
            catch (Exception ex)
            {
                throw;
            }
        }

        private string Max(Dictionary<string, double> dict)
        {
            string res = "";
            double max = -1d;

            foreach (var kv in dict)
            {
                if (kv.Value > max && kv.Value > 0.0d)
                {
                    max = kv.Value;
                    res = kv.Key;
                }
            }

            return res;
        }
    }
}

File: Awesome.AI.Python/Core/Spaces/HubSpace.py
"""Weighted hub-space port of ``Core/Spaces/HubSpace.cs``."""
import random

from awesome_ai.Core.Internals.Lookup import Lookup
from awesome_ai.Variables.Constants import CONST


class HubSpace:
    def __init__(self, mind):
        self.mind, self.weights = mind, {}
        lookup = Lookup()
        for occupation in CONST.occupasions:
            for hub in lookup.GetHUBS(mind.mindtype, occupation):
                if hub not in self.weights:
                    self.weights[hub] = random.random()

    def _occupation(self):
        return self.mind._internal.Occu.name

    def _occupation_hubs(self):
        return Lookup().GetHUBS(self.mind.mindtype, self._occupation())

    def AdjustWeights(self, sub, value):
        hubs = self._occupation_hubs()
        for hub in hubs:
            self.weights[hub] -= value * 0.1
        for hub in hubs:
            self.weights[hub] = 0.0 if self.weights[hub] < 0.0 else self.weights[hub]
        self.weights[sub] += value

    def GetIndex(self, subject):
        hubs = self._occupation_hubs()
        weights = [(hub, self.weights[hub]) for hub in hubs]
        total = sum(value for _, value in weights)
        count = 0.0
        result = 0.0
        for index, (hub, value) in enumerate(weights):
            count += value / total * 100.0
            if subject == self.GetSubject(count):
                break
            if index + 1 > len(weights) - 1:
                return -1.0
            next_area = weights[index + 1][1] / total * 100.0
            result = count + next_area / 2.0
        return result

    @staticmethod
    def _long_decision(unit):
        if not unit.IsDECISION():
            return ""
        data = unit.Data
        if data.startswith(CONST.lng_should):
            return CONST.LSUB_SHOULD
        if data.startswith(CONST.lng_what):
            return CONST.LSUB_WHAT
        return ""

    def GetSubject(self, value):
        if hasattr(value, "IsDECISION"):
            decision_subject = self._long_decision(value)
            if decision_subject:
                return decision_subject
            hub_index = value.HIget()
            if self._occupation() == "init":
                return "init"
        else:
            hub_index = float(value)

        hubs = self._occupation_hubs()
        # This intentionally uses every hub's weight, matching the source's
        # overloaded GetSubject implementation rather than only this occupation.
        total = sum(self.weights.values())
        if not hubs:
            return ""
        count = 0.0
        subject = ""
        for hub in hubs:
            count += self.weights[hub] / total * 100.0
            subject = hub
            if count >= hub_index:
                break
        return subject

    def UnitsPerHub(self, hub):
        return [unit for unit in self.mind.access.UNITS_ALL() if self.GetSubject(unit.HIget()) == hub]

    def UnitsPerOccupasionc(self):
        return [unit for unit in self.mind.access.UNITS_ALL() if unit.IsValid]

    @staticmethod
    def Max(values):
        result, maximum = "", -1.0
        for key, value in values.items():
            if value > maximum and value > 0.0:
                result, maximum = key, value
        return result
